ga4gh/backend.py

```diff
@@ -391,26 +392,150 @@ def _topLevelObjectGenerator(self, request, numObjects, getByIndexMethod):
                 nextPageToken = str(currentIndex)
             yield object_.toProtocolElement(), nextPageToken
 
-    def _objectListGenerator(self, request, objectList):
+    def _filterReadGroupsByBioSampleId(self, bioSampleId, readGroupSet):
         """
-        Returns a generator over the objects in the specified list using
-        _topLevelObjectGenerator to generate page tokens.
+        Accepts a request object containing a biosampleId and
+        removes read groups that don't match the given id.
+        :param str bioSampleId:
+        :param readGroupSet:
+        :return: readGroupSet
         """
-        return self._topLevelObjectGenerator(
-            request, len(objectList), lambda index: objectList[index])
+        newRg = []
+        for rg in readGroupSet.readGroups:
+            if bioSampleId == rg.bioSampleId:
+                newRg.append(rg)
+        readGroupSet.readGroups = newRg
+        return readGroupSet
 
-    def _singleObjectGenerator(self, datamodelObject):
+    def _readGroupSetObjectGenerator(
+            self, request, numObjects, getByIndexMethod, recur=True):
         """
-        Returns a generator suitable for a search method in which the
-        result set is a single object.
+        Returns read group sets that match the given request.
+        Filters read groups by a request's bioSampleId. In order
+        to properly set the nextPageToken the function searches
+        ahead to the next page to see if there are results using
+        a recursive call.
         """
-        yield (datamodelObject.toProtocolElement(), None)
+        currentIndex = 0
+        if request.pageToken is not None:
+            currentIndex, = _parsePageToken(request.pageToken, 1)
+        while currentIndex < numObjects:
+            readGroupSet = getByIndexMethod(currentIndex).toProtocolElement()
+            currentIndex += 1
+            nextPageToken = None
+            if currentIndex < numObjects and recur:
+                request.pageToken = str(currentIndex)
+                for rgs, pageToken in self._readGroupSetObjectGenerator(
+                        request, numObjects, getByIndexMethod, False):
+                    nextPageToken = str(currentIndex)
+                    break
+            if (request.name == "" or request.name) and request.bioSampleId:
+                if request.name == readGroupSet.name:
+                    yield self._filterReadGroupsByBioSampleId(
+                        request.bioSampleId, readGroupSet), nextPageToken
+            elif (request.name == "" or request.name):
```

Comment:
It's confusing what this is checking for.  Using a string as a boolean is a good thing to avoid  to keep code readable. Is this really a check for `is not None' ??